sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsMeta.scala

```diff
@@ -119,23 +119,9 @@ abstract class RapidsMeta[INPUT <: BASE, BASE, OUTPUT <: BASE](
 
   val gpuSupportedTag = TreeNodeTag[Set[String]]("rapids.gpu.supported")
 
-  /**
-   * Recursively force a section of the plan back onto CPU, stopping once a plan
-   * is reached that is already on CPU.
-   */
-  final def recursiveCostPreventsRunningOnGpu(): Unit = {
-    if (canThisBeReplaced) {
-      costPreventsRunningOnGpu()
-      childDataWriteCmds.foreach(_.recursiveCostPreventsRunningOnGpu())
-    }
-  }
-
   final def costPreventsRunningOnGpu(): Unit = {
     cannotRunOnGpuBecauseOfCost = true
     willNotWorkOnGpu("Removed by cost-based optimizer")
-    childExprs.foreach(_.recursiveCostPreventsRunningOnGpu())
-    childParts.foreach(_.recursiveCostPreventsRunningOnGpu())
-    childScans.foreach(_.recursiveCostPreventsRunningOnGpu())
```

Comment:
Why were these removed? are we not reporting anymore that they cannot run on the GPU because of cost?

Reply:
This code was originally based on the logic used in `recursiveSparkPlanPreventsRunningOnGpu` and `recursiveSparkPlanRemoved` but in other cases (like fixing up join consistency) we were just calling `willNotWorkOnGpu` without recursing into the expr, parts, and scans because we are just tagging that the operator should not run on GPU. 

I wasn't sure why we would want to do this recursion part other than recursing through the operators, and removing this code didn't affect the unit tests or the results when running benchmarks so I removed it. Is this something that should be added back in? If so, do you have any guidance on how to test this part in the context of the optimizer?

Reply:
The question really is about what is printed out when the GPU explain is enabled.  

Before this change all of the expressions, parts, and scans that are children would report that they were "Removed by cost-based optimizer" and include the `$` at the beginning indicating that it was removed because of cost.

After this change the expressions, parts, and scans will not say anything about the cost based optimizer, but will print out a `@` at the beginning and say it could run on the GPU but will not because of the spark plan it is a part of.

Which do we want to happen?  I don't know. Both appear reasonable to me. I just assumed that the code was put in on purpose but I guess not.

Reply:
Thanks. I think I should put this back in then. It does require adding another type of tag to RapidsMeta to track the concept of `mustThisBeReplaced` though.